src/pyclass.rs

```diff
@@ -71,185 +66,20 @@ pub unsafe fn tp_free_fallback(obj: *mut ffi::PyObject) {
-    const IS_NATIVE_TYPE: bool = false;
-    fn get_super_or(&mut self) -> Option<&mut <T::BaseType as PyTypeInfo>::ConcreteLayout> {
-        Some(&mut self.ob_base)
-    }
-    unsafe fn internal_ref_cast(obj: &PyAny) -> &T {
-        let shell = obj.as_ptr() as *const Self;
-        &(*shell).pyclass
-    }
-    unsafe fn internal_mut_cast(obj: &PyAny) -> &mut T {
-        let shell = obj.as_ptr() as *const _ as *mut Self;
-        &mut (*shell).pyclass
-    }
-    unsafe fn py_drop(&mut self, py: Python) {
-        ManuallyDrop::drop(&mut self.pyclass);
-        self.dict.clear_dict(py);
-        self.weakref.clear_weakrefs(self.as_ptr(), py);
-        self.ob_base.py_drop(py);
-    }
-    unsafe fn py_init(&mut self, value: T) {
-        self.pyclass = ManuallyDrop::new(value);
-    }
-}
-
-impl<T: PyClass> PyObjectSizedLayout<T> for PyClassShell<T> {}
-
-impl<T: PyClass> AsPyPointer for PyClassShell<T> {
-    fn as_ptr(&self) -> *mut ffi::PyObject {
-        (self as *const _) as *mut _
-    }
-}
-
-impl<T: PyClass> std::ops::Deref for PyClassShell<T> {
-    type Target = T;
-    fn deref(&self) -> &T {
-        self.pyclass.deref()
-    }
-}
-
-impl<T: PyClass> std::ops::DerefMut for PyClassShell<T> {
-    fn deref_mut(&mut self) -> &mut T {
-        self.pyclass.deref_mut()
-    }
-}
-
-impl<T: PyClass> ToPyObject for &PyClassShell<T> {
-    fn to_object(&self, py: Python<'_>) -> PyObject {
-        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
-    }
-}
-
-impl<T: PyClass> ToPyObject for &mut PyClassShell<T> {
-    fn to_object(&self, py: Python<'_>) -> PyObject {
-        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
-    }
-}
-
-unsafe impl<'p, T> FromPyPointer<'p> for &'p PyClassShell<T>
-where
-    T: PyClass,
-{
-    unsafe fn from_owned_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
-        NonNull::new(ptr).map(|p| &*(gil::register_owned(py, p).as_ptr() as *const PyClassShell<T>))
-    }
-    unsafe fn from_borrowed_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
-        NonNull::new(ptr)
-            .map(|p| &*(gil::register_borrowed(py, p).as_ptr() as *const PyClassShell<T>))
-    }
-}
-
-unsafe impl<'p, T> FromPyPointer<'p> for &'p mut PyClassShell<T>
-where
-    T: PyClass,
-{
-    unsafe fn from_owned_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
-        NonNull::new(ptr).map(|p| {
-            &mut *(gil::register_owned(py, p).as_ptr() as *const _ as *mut PyClassShell<T>)
-        })
-    }
-    unsafe fn from_borrowed_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
-        NonNull::new(ptr).map(|p| {
-            &mut *(gil::register_borrowed(py, p).as_ptr() as *const _ as *mut PyClassShell<T>)
-        })
-    }
+    /// The closest native ancestor. This is `PyAny` by default, and when you declare
+    /// `#[pyclass(extends=PyList)]`, it's `PyList`.
+    type BaseNativeType: PyTypeInfo + PyNativeType;
```

Comment:
I thought we can't safely extend from `PyTuple` or `PyList` yet, so might want to change this doc?

Reply:
Oops, good catch, thanks.